drivers/can/can_nxp_s32_canxl.c

```diff
@@ -0,0 +1,1020 @@
+		if (frame->dlc > CANFD_MAX_DLC) {
+			LOG_ERR("DLC of %d for CAN-FD format frame", frame->dlc);
+			return -EINVAL;
+		}
+	} else {
+		if (frame->dlc > CAN_MAX_DLC) {
+			LOG_ERR("DLC of %d for non-FD format frame", frame->dlc);
+			return -EINVAL;
+		}
+	}
+
+	if (!data->started) {
+		return -ENETDOWN;
+	}
+
+	can_nxp_s32_get_state(dev, &state, NULL);
+	if (state == CAN_STATE_BUS_OFF) {
+		LOG_ERR("Transmit failed, bus-off");
+		return -ENETUNREACH;
+	}
+
+	if (k_sem_take(&data->tx_allocs_sem, timeout) != 0) {
+		return -EAGAIN;
+	}
+
+	for (alloc = 0; alloc < CAN_NXP_S32_MAX_TX; alloc++) {
+		if (!atomic_test_and_set_bit(data->tx_allocs, alloc)) {
+			break;
+		}
+	}
+
+	data->tx_cbs[alloc].function = callback;
+	data->tx_cbs[alloc].arg = user_data;
+	mb_indx = ALLOC_IDX_TO_TXMB_IDX(alloc);
+	data->tx_cbs[alloc].tx_info = (Canexcel_Ip_DataInfoType) {
+		.frame = !!(frame->flags & CAN_FRAME_FDF) ?
+				CANEXCEL_FD_FRAME : CANEXCEL_CLASIC_FRAME,
+		.enable_brs = !!(frame->flags & CAN_FRAME_BRS) ? TRUE : FALSE,
+		.idType = !!(frame->flags & CAN_FRAME_IDE) ?
+				CANEXCEL_MSG_ID_EXT : CANEXCEL_MSG_ID_STD,
+		.priority = 0,
+		.retransmission = 0,
+		.fd_padding = 0,
+		.dataLength = data_length,
+		.is_polling = FALSE
+	};
+
+	LOG_DBG("%s: Sending %d bytes Tx Mb %d, "
+		"Tx Id: 0x%x, "
+		"Id type: %s %s %s %s"
+		, dev->name, data_length
+		, mb_indx, frame->id
+		, !!(frame->flags & CAN_FRAME_IDE) ?
+						"extended" : "standard"
+		, !!(frame->flags & CAN_FRAME_RTR) ? "RTR" : ""
+		, !!(frame->flags & CAN_FRAME_FDF) ? "FD frame" : ""
+		, !!(frame->flags & CAN_FRAME_BRS) ? "BRS" : "");
+
+	k_mutex_lock(&data->tx_mutex, K_FOREVER);
+	config->irq_disable_func();
+	/* Send MB Interrupt */
+	status = Canexcel_Ip_SendFDMsg(config->instance, mb_indx, &data->tx_cbs[alloc].tx_info,
+				frame->id, (uint8_t *)&frame->data, &data->tx_msg[alloc]);
+	config->irq_enable_func();
+	k_mutex_unlock(&data->tx_mutex);
+
+	if (status != CANEXCEL_STATUS_SUCCESS) {
+		return -EIO;
+	}
+
+	return 0;
+}
+
+static void nxp_s32_zcan_timming_to_canxl_timming(const struct can_timing *timing,
+							Canexcel_Ip_TimeSegmentType *canxl_timing)
+{
+	LOG_DBG("propSeg: %d, phase_seg1: %d, phase_seg2: %d, prescaler: %d, sjw: %d",
+		timing->prop_seg, timing->phase_seg1, timing->phase_seg2,
+		timing->prescaler, timing->sjw);
+
+	canxl_timing->propSeg = timing->prop_seg - 1U;
+	canxl_timing->phaseSeg1 = timing->phase_seg1 - 1U;
+	canxl_timing->phaseSeg2 = timing->phase_seg2 - 1U;
+	canxl_timing->preDivider = timing->prescaler - 1U;
+	canxl_timing->rJumpwidth = timing->sjw - 1U;
+}
+
+static int can_nxp_s32_set_timing(const struct device *dev,
+			const struct can_timing *timing)
+{
+	const struct can_nxp_s32_config *config = dev->config;
+	struct can_nxp_s32_data *data = dev->data;
+	Canexcel_Ip_TimeSegmentType can_time_segment = {0};
+
+	if (data->started) {
+		return -EBUSY;
+	}
+
+	LOG_DBG("Setting CAN bitrate %d:", config->bitrate);
```

Comment:
This debug log looks wrong. It uses the bitrate species in DTS instead of the timing parameters supplied to the function.